File: testdata/precision-suite/tp_ssti.go
// Server-side template injection: a request-controlled string is parsed as a
// text/template and executed, letting an attacker inject template actions —
// CWE-1336 / CWE-94. A correct scanner fires TAINT-003 (the taint engine's
// SSTI sink). Unlike the Python SSTI sample (caught by the VARIANT-005 CVE
// signature), nox has no Go variant signature and no Go taint model, so this
// is an honest recall gap.
package render

import (
	"net/http"
	"text/template"
)

// greet parses a caller-supplied template string, then executes it — the vulnerability.
func greet(w http.ResponseWriter, r *http.Request) {
	src := r.URL.Query().Get("tmpl")
	tmpl, err := template.New("greeting").Parse(src) // nox-expect: TAINT-003
	if err != nil {
		http.Error(w, "bad template", http.StatusBadRequest)
		return
	}
	_ = tmpl.Execute(w, map[string]string{"name": "world"})
}

File: testdata/precision-suite/tp_ssti.py
# Server-side template injection: user input concatenated into a rendered
# template. nox catches this via the VARIANT-005 signature (dynamic template
# string), so it is a true positive — but see the density view: it should fire
# exactly once here.
from flask import render_template_string  # nox-expect: SLOP-001
def greet(request):
    name = request.args.get("name")
    return render_template_string("<h1>Hello " + name + "</h1>")  # nox-expect: VARIANT-005
